File: 2016/luke12.hs
import Data.Maybe
import Data.Char

main :: IO ()
main = do content <- lines <$> readFile "luke12_data.txt"
          let decs   = fromEng <$> (filter (`elem` ['a'..'z']) . map toLower $ head content)
              pair d = let half = d `div` 2
                       in if odd d then (half + 1, half) else (half, half)
              paired = pair <$> decs
              coded  = (fst <$> paired) ++ reverse (snd <$> paired)

          print $ fromDec <$> coded

fromEng :: Char -> Integer
fromEng n = fromJust (lookup n (zip ['a'..'z'] [1..26]))

fromDec :: Integer -> String
fromDec d = fromMaybe "0" (lookup d (zip [1..13] ["I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII"]))

File: 2016/luke12.rb
from_eng = Hash[('a'..'z').zip(1..26)]
from_dec = Hash[(0..13).zip(%w(0 I II III IV V VI VII VIII IX X XI XII XIII))]

decs = File.readlines('luke12_data.txt').first.strip.downcase.split(//)
  .select { |e| ('a'..'z').include? e }
  .map { |e| from_eng[e] }
  .map { |d| if d.odd? then [d.div(2) + 1, d.div(2)] else [d.div(2), d.div(2)] end }

p ((decs.map &:first) + (decs.map &:last).reverse)
  .map { |i| from_dec[i] }.join(",")
